### ISSUE TITLE
Input stream line position and read term start and end lines

### ISSUE THREAD

**Issue by pmoura** [tau-prolog/tau-prolog#152]
As a future enhancement, it would be quite useful if we could call the `stream_property/2` predicate to get the current line and if the `read_term/3` predicate could support an option to get the term start and ending lines. In my particular case, Logtalk uses these features to pinpoint warnings and errors when compiling a file. When we use the sandbox in the Tau Prolog website, warnings and errors tell us the source file line number. Thus, I assume the some of the necessary bits to support these features are already present internally?

**Comment by jariazavalverde** [comment]
Yes, currently the tokens store that information. What is the convention for this property? For the same stream, SWI-Prolog gives me `'$stream_position'(57,1,57,57)` and GNU Prolog `'$stream_position'(57,57,0,57)`.

**Comment by pmoura** [comment]
From the ISO Prolog standard perspective, the `position/1` property value is not specified. You can regard it as an opaque term and then have a small set of predicates to access the information in there. This works well and enables changing position information without breaking existing code. On the other hand, there's not that much to tell about a stream position. We have line numbers and possibly column numbers for text streams. Thus, you can always go with a simple term like `position(Line,column)` for the `stream_property/2` predicate `position/1` property. For `read_term/3`, ideally we have not just the start position or the end position of the term but both.

**Comment by pmoura** [comment]
Btw, you can look inside the `logtalk3/adapters` files and see how different Prolog systems do it (search the files for the predicates `'$lgt_stream_current_line_number'` and `'$lgt_read_term'`.

**Comment by jariazavalverde** [comment]
Added `stream_position_data/3` built-in predicate.

`stream_position_data(?Field, +Pos, -Data)`. `Field` is one of `line_count`, `line_position` or `char_count`.

**Comment by pmoura** [comment]
There seems to be an initialization bug when reading the first term from a stream:

```text
.../node_modules/tau-prolog/modules/core.js:7848
					if(last_token.line_position === last_token.len)
					              ^

TypeError: Cannot read property 'line_position' of null
    at read_term/3 (.../node_modules/tau-prolog/modules/core.js:7848:20)
```

**Comment by jariazavalverde** [comment]
Thanks. Fixed!

**Comment by pmoura** [comment]
Looking into the implementation of the `read_term/3` predicate, it seems like all the necessary bits to support a `term_position(Position)` option are there. This option would return the same opaque term  returned by the `stream_property(Stream, position(Position))`. Worthwhile if `Position` would be the start of the term not counting any whitespace or comments that precede it (something that is not possible by calling the `stream_property/2` predicate).
